File: differ/src/main/java/ru/starksoft/differ/viewholder/DifferViewHolder.java
package ru.starksoft.differ.viewholder;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.ref.WeakReference;

import androidx.annotation.CallSuper;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import ru.starksoft.differ.adapter.OnClickListener;
import ru.starksoft.differ.viewmodel.ViewModel;

import static androidx.recyclerview.widget.RecyclerView.NO_POSITION;

public abstract class DifferViewHolder <M extends ViewModel> extends RecyclerView.ViewHolder {

	@Nullable private final WeakReference<OnClickListener> adapterWeakReference;
	private M viewModel;

	public DifferViewHolder(@LayoutRes int layout, @NonNull ViewGroup parent, @Nullable OnClickListener onClickListener) {
		super(LayoutInflater.from(parent.getContext()).inflate(layout, parent, false));
		setBinder(itemView);
		if (onClickListener != null) {
			adapterWeakReference = new WeakReference<>(onClickListener);
		} else {
			adapterWeakReference = null;
		}
	}

	protected abstract void setBinder(@NonNull View view);

	@CallSuper
	public void bind(@NonNull M viewModel, @Nullable Bundle diff) {
		this.viewModel = viewModel;
		if (diff != null) {
			bindPayloads(diff);
		} else {
			bind(viewModel);
		}
	}

	/**
	 * Полное обновление данных, в том числе через databinding
	 *
	 * @param viewModel ViewModel
	 */
	protected void bind(@NonNull M viewModel) {
	}

	/**
	 * Частичное обновление данных
	 *
	 * @param diff Bundle с данными требующимии обновления на view'шке
	 */
	protected void bindPayloads(@NonNull Bundle diff) {
	}

	@NonNull
	public M getViewModel() {
		return viewModel;
	}

	public void onClick(int action) {
		onClick(action, new Bundle());
	}

	public void onClick(int action, @NonNull Bundle extra) {
		if (adapterWeakReference != null) {
			OnClickListener listener = adapterWeakReference.get();
			if (listener != null) {
				int position = getAdapterPosition();
				if (position != NO_POSITION) {
					listener.onClick(getAdapterPosition(), getViewModel(), action, extra);
				}
			}
		}

	}

	/**
	 * RecyclerView calls this method right before clearing ViewHolder's internal data and
	 * sending it to RecycledViewPool
	 */
	public void onUnbind() {
	}
}

File: diff-adapter/src/main/java/ru/starksoft/differ/adapter/viewholder/DifferViewHolder.kt
package ru.starksoft.differ.adapter.viewholder

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.NO_POSITION
import ru.starksoft.differ.adapter.OnClickListener
import ru.starksoft.differ.adapter.viewmodel.ViewModel
import java.lang.ref.WeakReference

abstract class DifferViewHolder<M : ViewModel>(@LayoutRes layout: Int, parent: ViewGroup, onClickListener: OnClickListener?) :
	RecyclerView.ViewHolder(LayoutInflater.from(parent.context).inflate(layout, parent, false)) {

	private val adapterWeakReference = if (onClickListener != null) WeakReference(onClickListener) else null

	lateinit var viewModel: M
		private set

	init {
		this.setBinder(itemView)
	}

	protected open fun setBinder(view: View) {}

	@CallSuper
	fun bind(viewModel: M, payload: Bundle?) {
		this.viewModel = viewModel
		payload?.let { bindPayloads(it) } ?: bind(viewModel)
	}

	/**
	 * Полное обновление данных, в том числе через databinding
	 *
	 * @param viewModel ViewModel
	 */
	protected open fun bind(viewModel: M) {}

	/**
	 * Частичное обновление данных
	 *
	 * @param payload Bundle с данными требующимии обновления на view'шке
	 */
	protected open fun bindPayloads(payload: Bundle) {}

	@JvmOverloads
	fun onClick(action: Int, extra: Bundle = Bundle()) {
		val position = adapterPosition
		adapterWeakReference?.get()?.let {
			if (position != NO_POSITION) {
				it.onClick(position, viewModel, action, extra)
			}
		}
	}

	/**
	 * RecyclerView calls this method right before clearing ViewHolder's internal data and
	 * sending it to RecycledViewPool
	 */
	open fun onUnbind() {}
}

